src/TwigRendererFactory.php

```diff
@@ -12,145 +12,25 @@
-            ));
-        }
-
-        $debug    = array_key_exists('debug', $config) ? (bool) $config['debug'] : false;
-        $config   = $this->mergeConfig($config);
-        $cacheDir = isset($config['cache_dir']) ? $config['cache_dir'] : false;
-
-        // Create the engine instance
-        $loader      = new TwigLoader();
-        $environment = new TwigEnvironment($loader, [
-            'cache'            => $debug ? false : $cacheDir,
-            'debug'            => $debug,
-            'strict_variables' => $debug,
-            'auto_reload'      => $debug
-        ]);
-
-        // Add expressive twig extension
-        if ($container->has(ServerUrlHelper::class) && $container->has(UrlHelper::class)) {
-            $environment->addExtension(new TwigExtension(
-                $container->get(ServerUrlHelper::class),
-                $container->get(UrlHelper::class),
-                isset($config['assets_url']) ? $config['assets_url'] : '',
-                isset($config['assets_version']) ? $config['assets_version'] : '',
-                isset($config['globals']) ? $config['globals'] : []
-            ));
-        }
-
-        // Add debug extension
-        if ($debug) {
-            $environment->addExtension(new TwigExtensionDebug());
-        }
-
-        // Add user defined extensions
-        $extensions = (isset($config['extensions']) && is_array($config['extensions']))
-            ? $config['extensions']
-            : [];
-        $this->injectExtensions($environment, $container, $extensions);
-
-        // Inject environment
-        $twig = new TwigRenderer($environment, isset($config['extension']) ? $config['extension'] : 'html.twig');
-
-        // Add template paths
-        $allPaths = isset($config['paths']) && is_array($config['paths']) ? $config['paths'] : [];
-        foreach ($allPaths as $namespace => $paths) {
-            $namespace = is_numeric($namespace) ? null : $namespace;
-            foreach ((array) $paths as $path) {
-                $twig->addPath($path, $namespace);
-            }
-        }
-
-        return $twig;
-    }
-
-    /**
-     * Inject extensions into the TwigEnvironment instance.
-     *
-     * @param TwigEnvironment $environment
-     * @param ContainerInterface $container
-     * @param array $extensions
-     * @throws Exception\InvalidExtensionException
-     */
-    private function injectExtensions(TwigEnvironment $environment, ContainerInterface $container, array $extensions)
-    {
-        foreach ($extensions as $extension) {
-            // Load the extension from the container
-            if (is_string($extension) && $container->has($extension)) {
-                $extension = $container->get($extension);
-            }
+        $config      = $container->has('config') ? $container->get('config') : [];
+        $config      = $this->mergeConfig($config);
+        $environment = $container->get(TwigEnvironment::class);
```

Comment:
Hm... This may be a BC break, on further review.

If a user was previously using the `TwigRenderer` service and upgrades, the `TwigEnvironment` service is not registered, which will cause this factory to fail.

My thought is to do this:

```php
$environment = $this->getEnvironment($container);
```

Where `getEnvironment()` is defined as follows:

```php
public function getEnvironment(ContainerInterface $container)
{
    if ($container->has(TwigEnvironment::class)) {
        return $container->get(TwigEnvironment::class);
    }

    $environmentFactory = new TwigEnvironmentFactory();
    return $environmentFactory($container);
}
```

This will keep usage backwards compatible for existing users, while allowing them to add the factory to their configuration in order to be forwards compatible.

I can do that on merge.